src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SslStreamSecurityUpgradeProvider.cs

```diff
@@ -566,16 +414,210 @@ internal override void Close(TimeSpan timeout)
         protected override Stream OnInitiateUpgrade(Stream stream, out SecurityMessageProperty remoteSecurity)
         {
             OutWrapper<SecurityMessageProperty> remoteSecurityWrapper = new OutWrapper<SecurityMessageProperty>(); 
-            Stream retVal = this.OnInitiateUpgradeAsync(stream, remoteSecurityWrapper).GetAwaiter().GetResult();
+            Stream retVal = OnInitiateUpgradeAsync(stream, remoteSecurityWrapper).GetAwaiter().GetResult();
             remoteSecurity = remoteSecurityWrapper.Value; 
 
             return retVal; 
         }
 
 #if FEATURE_NETNATIVE
-        protected override Task<Stream> OnInitiateUpgradeAsync(Stream stream, OutWrapper<SecurityMessageProperty> remoteSecurityWrapper)
+        protected override async Task<Stream> OnInitiateUpgradeAsync(Stream stream, OutWrapper<SecurityMessageProperty> remoteSecurityWrapper)
         {
-            throw ExceptionHelper.PlatformNotSupported("SslStreamSecurityUpgradeInitiator.InInitiateUpgradeAsync");
+            if (WcfEventSource.Instance.SslOnInitiateUpgradeIsEnabled())
+            {
+                WcfEventSource.Instance.SslOnInitiateUpgrade();
+            }
+
+            // There is currently no way to convert a .Net X509Certificate2 to a UWP Certificate. The client certificate
+            // needs to be provided by looking it up in the certificate store. E.g.
+            //
+            //     factory.Credentials.ClientCertificate.SetCertificate(
+            //         StoreLocation.CurrentUser,
+            //         StoreName.My,
+            //         X509FindType.FindByThumbprint,
+            //         clientCertThumb);
+            //
+            // The certificate is retrieved using .Net api's and UWP api's. An artifical X509Extension is used to attach the
+            // UWP certificate to the .Net X509Certificate2. This is then retrieved at the point of usage to use with UWP
+            // networking api's.
+
+            Certificate clientCertificate = null;
+
+            if (_clientToken != null)
+            {
+                foreach (var extension in _clientToken.Certificate.Extensions)
+                {
+                    var attachmentExtension =
+                        extension as X509CertificateInitiatorClientCredential.X509UwpCertificateAttachmentExtension;
+                    if (attachmentExtension != null && attachmentExtension.AttachedCertificate != null)
+                    {
+                        clientCertificate = attachmentExtension.AttachedCertificate;
+                        break;
+                    }
+                }
+            }
+
+            Contract.Assert(clientCertificate != null, "Missing UWP Certificate as an attachment to X509Certificate2");
```

Comment:
For UWP apps, we don't allow directly setting an X509Certificate2, the only allowed way is using the SetCertificate method to fetch it from the store. If an X509Certificate2 is found, then we expect a UWP Certificate to also be found and throw if that's not the case. So the only way we can get an X509Certificate2 is if an UWP Certificate has also been found and attached.